components/suggest/src/yelp.rs

```diff
@@ -0,0 +1,325 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+use interrupt_support::SqlInterruptScope;
+use rusqlite::types::ToSqlOutput;
+use rusqlite::{named_params, Connection, Result as RusqliteResult, ToSql};
+use sql_support::ConnExt;
+use url::Url;
+
+use crate::{
+    provider::SuggestionProvider, rs::DownloadedYelpDataset, suggestion::Suggestion, Result,
+    SuggestionQuery,
+};
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+#[repr(u8)]
+enum Modifier {
+    PreModifier = 0,
+    PostModifier = 1,
+}
+
+impl ToSql for Modifier {
+    fn to_sql(&self) -> RusqliteResult<ToSqlOutput<'_>> {
+        Ok(ToSqlOutput::from(*self as u8))
+    }
+}
+
+/// Insert Yelp dataset into db.
+pub(crate) fn insert_dataset(
+    conn: &Connection,
+    scope: &SqlInterruptScope,
+    dataset: &DownloadedYelpDataset,
+) -> Result<()> {
+    for subject in &dataset.subjects {
+        scope.err_if_interrupted()?;
+        conn.execute("INSERT INTO yelp_subjects(value) VALUES(?)", [subject])?;
+    }
+
+    for modifier in &dataset.pre_modifiers {
+        scope.err_if_interrupted()?;
+        conn.execute(
+            "INSERT INTO yelp_modifiers(type, value) VALUES(:type, :modifier)",
+            named_params! {
+                ":type": Modifier::PreModifier,
+                ":modifier": modifier,
+            },
+        )?;
+    }
+
+    for modifier in &dataset.post_modifiers {
+        scope.err_if_interrupted()?;
+        conn.execute(
+            "INSERT INTO yelp_modifiers(type, value) VALUES(:type, :modifier)",
+            named_params! {
+                ":type": Modifier::PostModifier,
+                ":modifier": modifier,
+            },
+        )?;
+    }
+
+    for location_sign in &dataset.location_signs {
+        scope.err_if_interrupted()?;
+        conn.execute(
+            "INSERT INTO yelp_location_signs(value) VALUES(?)",
+            [location_sign],
+        )?;
+    }
+
+    scope.err_if_interrupted()?;
+    conn.execute(
+        "INSERT INTO yelp_templates(url) VALUES(?)",
+        [&dataset.template],
+    )?;
+
+    Ok(())
+}
+
+/// Fetch Yelp suggestion from given user's query.
+pub(crate) fn fetch_suggestion(
+    conn: &Connection,
+    query: &SuggestionQuery,
+) -> Result<Option<Suggestion>> {
+    if !query.providers.contains(&SuggestionProvider::Yelp) {
+        return Ok(None);
+    }
+
+    if !&query.keyword.contains(" ") {
```

Comment:
Yeah, I agree!
At this time, the suggestions module can provide several suggestions including Yelp, Wikipedia.
And, the limitation might change by the suggestion. For example, Yelp can accept 100 chars, Wikipedia  255 chars etc. If so, the consumer can't restrict since does not know which suggestion will be returned by the user's query on the consumer side. And I think it would be more independent. So, I feel we can restrict here.

On a similar note, and related to the Yelp title, I feel that it might be better not to make the query lowercase in the consumer. As much as possible, why not just pass the query as it is? What do you think??
https://searchfox.org/mozilla-central/rev/16526d690cccc9c60cacf09cc08e2c3041ef884e/browser/components/urlbar/private/SuggestBackendRust.sys.mjs#100

Reply:
I agree with both of your comments (restricting here in the Rust component and not lowercasing on desktop). I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1868922 for the desktop part.